Basic text model

janmalyy opened the pull request:
Please look at the lemmatize_text_data_WIP.py script, there is an issue I can not resolve (marked TODO) and hopes that if someone else look at it, he will find the solution

Commit:
improved preprocess_data script by wrapping into functions and main

Commit:
added .gitignore

Commit:
worked on lemmatization WIP

Commit:
removed .idea

janrygl commented on lemmatize_text_data_WIP.py:
```diff
@@ -0,0 +1,160 @@
+    creates a dataframe with lemmatized texts
+    takes long time to execute!!! (around 20 minutes) because of the calls to external API with all text data
+    it is split into so many parts because we want to avoid the HTTP run out of the time exception
+    """
```
it is better to use lemmatization library instead of API call (isn't Prague lemmatizer free?)
it is good to cache results not to need to run it again and again
e.g. make JSON with already translated sentences

janrygl replied:
I suggest to use pymorphodita, they have better models and tools than NLP lab

Not sure if it works, but there is a chatGPT example. I suggest to make lemmatizer module and import it as needed:

```
import pymorphodita

# Load the pre-trained Czech model
def load_morphodita_model(model_path):
    tagger = pymorphodita.Tagger.load(model_path)
    if not tagger:
        raise Exception("Failed to load the MorphoDiTa model")
    return tagger

# Lemmatize the input Czech text
def lemmatize_text(tagger, text):
    tokenizer = pymorphodita.Tokenizer.new_utf8_tokenizer()
    tokenizer.set_text(text)
    
    lemmas = []
    forms = pymorphodita.Forms()
    lemmas_list = pymorphodita.TaggedLemmas()
    
    while tokenizer.next_sentence(forms):
        tagger.tag(forms, lemmas_list)
        for lemma in lemmas_list:
            lemmas.append(lemma.lemma.split('_')[0])  # Extracting the base lemma
    return lemmas

# Example usage
if __name__ == "__main__":
    model_path = 'czech-morfflex-pdt-161115.tagger'  # Path to the MorphoDiTa model
    tagger = load_morphodita_model(model_path)
    
    text = "Byl jsem včera na návštěvě u babičky."  # Example Czech sentence
    lemmas = lemmatize_text(tagger, text)
    
    print("Lemmas:", lemmas)
```

janrygl replied:
the code above is crap, just example of morphodita using
It uses line comments instead of docstrings, it is not using default value for model path, etc.

otrnovec replied:
I wasn't able to find/install pymorphodita specifically but I ahve used ufal.morphodita which I believe should be basically the same.

Commit:
created very basic text model

Commit:
added stop words removal

Commit:
added a comment to make it more clear

otrnovec committed:
morphodita for lemmatizing + dataset is lemmatized

Commit:
deleted old dataset
tried some improvements in basic text model

janrygl commented on __pycache__/preprocess_data.cpython-39.pyc:
*pyc files and __pycache__ dirs should be in `.gitignore` not to be added to git

janrygl commented on basic_text_model_WIP.py:
```diff
@@ -0,0 +1,89 @@
+import matplotlib.pyplot as plt
+import numpy as np
+
+czech_stop_list = [
```
can be reused later by other modules, make it a separate importable file?

If this is constant, I suggest to use upper casing: `CZECH_STOPWORDS = `

janrygl commented on basic_text_model_WIP.py:
```diff
@@ -0,0 +1,89 @@
+]
+
+
+def remove_stop_words(input_string: str, stop_list: list = czech_stop_list) -> str:
```
list[str]

janrygl commented on basic_text_model_WIP.py:
```diff
@@ -0,0 +1,89 @@
+def remove_stop_words(input_string: str, stop_list: list = czech_stop_list) -> str:
+    words = input_string.split()
+    filtered_words = [word for word in words if word not in stop_list]
+    output_string = ' '.join(filtered_words)
```
It is not  a problem, but you also convert \n newlines to spaces in this code. Document it, or work with tokenized text.

janrygl commented on basic_text_model_WIP.py:
```diff
@@ -0,0 +1,89 @@
+    return output_string
+
+
+def load_and_split_data() -> tuple:
```
tuple of what? I am just guessing:
`tuple[list[str], list[int], list[str], list[int]`

Dict is better for complex output, e.g.
`{"X_train": X_train, ...`

janrygl commented on basic_text_model_WIP.py:
```diff
@@ -0,0 +1,89 @@
+
+
+def load_and_split_data() -> tuple:
+    lemmatized_df = pd.read_csv(Path(__file__).parent / "data" / "lemmatized_dataset.csv")
```
what if paths don't exist?
it is better to define DATA_PATH once and then reuse

janrygl commented on basic_text_model_WIP.py:
```diff
@@ -0,0 +1,89 @@
+    df_train["status"] = df_train["status"].astype("int")
+    df_val["status"] = df_val["status"].astype("int")
+
+    # text columns are joint together because CountVectorizer accepts only one string column
```
good approach, but you can try also models using only one column and compare them

janrygl commented on lemmatize_text_data_WIP.py:
```diff
@@ -0,0 +1,43 @@
+from settings import DATA_DIR, MODEL_DIR
```
We usually do not put WIP into file name.
If we want to suggest WIP, we put to the function in progress:
 - raise NotImplementedError
 -  or # FIXME comment

janrygl commented on lemmatize_text_data_WIP.py:
```diff
@@ -0,0 +1,43 @@
+        }
+    )
+
+    df["status"].replace(
```
this mapping could be put to separate file to be easily founded -- and even README can link that file

otrnovec replied:
do you mean a different file than settings.py?

janrygl commented on lemmatize_text_data_WIP.py:
```diff
@@ -0,0 +1,43 @@
+        inplace=True,
+    )
+
+    df.to_csv(DATA_DIR / "lemmatized_dataset.csv", index=False)
```
I am missing return function.
Because `get_something` function suggests this.
This function should be named `generate_lemmatized_df(input, output)`.

janrygl commented on settings.py:
```diff
@@ -0,0 +1,6 @@
+
+BASE_DIR = Path(__file__).parent
+DATA_DIR = BASE_DIR / "data"
+MODEL_DIR = BASE_DIR / "czech-morfflex2.0-pdtc1.0-220710"
```
is this a model dir or a model file?

otrnovec replied:
its a model dir, changing the name

janrygl reviewed (approved):
Nice improvement. I didn't read everything, but it looks as a great start.

otrnovec committed:
resolving comments, proveditelný converting to 1

Commit:
fixed dataset balancing

Commit:
fixed dataset balancing

Commit:
retyped str>int in search in status column

otrnovec committed:
WIP word embeddings

otrnovec committed:
Merge branch 'basic-text-model' of https://github.com/otrnovec/PorcupAIne into basic-text-model

otrnovec committed:
-renamed basic_text_model (removed WIP)
-word_embeddigns should be done (im putting the embedded dataset on google drive)

Commit:
created text_model using word embeddings as features

janmalyy committed:
Merge branch 'main' into basic-text-model
